charts/core/src/Echart/useEchart.ts

```diff
@@ -0,0 +1,386 @@
+    withInstanceCheck((action, callback) => {
+      switch (action) {
+        case 'zoomselect': {
+          echartsInstance.off('datazoom', callback);
+          // Remove from active handlers
+          activeHandlers.current.delete(`${action}-${callback.toString()}`);
+          break;
+        }
+
+        default: {
+          echartsInstance.off(action, callback);
+          activeHandlers.current.delete(`${action}-${callback.toString()}`);
+        }
+      }
+    }),
+    [echartsInstance],
+  );
+
+  const on: EChartsInstance['on'] = useCallback(
+    withInstanceCheck((action, callback) => {
+      // Create a unique key for this handler
+      const handlerKey = `${action}-${callback.toString()}`;
+
+      // If this exact handler is already registered, skip
+      if (activeHandlers.current.has(handlerKey)) {
+        return;
+      }
+
+      switch (action) {
+        case 'zoomselect': {
+          const zoomHandler = (params: any) => {
+            const zoomSelectionEvent: ZoomSelectionEvent = {};
+
+            if (
+              params.startValue !== undefined &&
+              params.endValue !== undefined
+            ) {
+              // Handle single axis zoom
+              const axis = params.dataZoomId?.includes('y') ? 'yAxis' : 'xAxis';
+              zoomSelectionEvent[axis] = {
+                startValue: params.startValue,
+                endValue: params.endValue,
+              };
+              callback(zoomSelectionEvent);
+            } else if (params.batch) {
+              // Handle batch zoom (multiple axes)
+              params.batch.forEach((batchItem: any) => {
+                if (batchItem.dataZoomId?.includes('y')) {
+                  zoomSelectionEvent.yAxis = {
+                    startValue: batchItem.startValue,
+                    endValue: batchItem.endValue,
+                  };
+                } else {
+                  zoomSelectionEvent.xAxis = {
+                    startValue: batchItem.startValue,
+                    endValue: batchItem.endValue,
+                  };
+                }
+              });
+              callback(zoomSelectionEvent);
+            }
+          };
+
+          // Store the wrapper function so we can remove it later
+          activeHandlers.current.set(handlerKey, zoomHandler);
+          echartsInstance.on('datazoom', zoomHandler);
+          break;
+        }
+
+        default: {
+          activeHandlers.current.set(handlerKey, callback);
+          echartsInstance.on(action, callback);
+        }
+      }
+    }),
+    [echartsInstance],
+  );
+
+  useEffect(() => {
+    (async function setupChart() {
+      try {
+        setError(null);
+
+        await initializeEcharts();
+
+        if (container) {
+          const newChart = echartsCore.init(container);
+          newChart.setOption(options);
+          window.addEventListener('resize', resizeHandler);
+          setEchartsInstance(newChart);
+          setReady(true);
+        }
```

Comment:
The functional programmer in me feels like `initializeEcharts` should return the modules, and then instead of module-level singletons, we store the modules as `refs` within the hook itself:

```ts
initializeEcharts.then(({ echartsCore }) => {
  echartsCoreModuleRef.current = echartsCore;
  ...

}).catch((err) => {
  // secondary benefit is that we can avoid the try/catch
})
```

Reply:
We could avoid the try/catch by just not using async/await here as well. I tend to be a fan of async/await over nesting as I just find it more readable, but it's just personal preference. 

I do agree about returning modules here though! However, really only the core library needs to be returned, as the remaining modules are only used internally. I'll refactor this and update it use chaining instead if that's our preference.

Reply:
async/await is fine. I was more focused on what's returned